### PR TITLE
Add client retries

### LINKED ISSUE #44: Add retries to the Icepeak client
The icepeak client does currently not support retrying. We worked around that in JM by adding retries ourselves, but it would be nicer and more logical if the client could do that itself.

This could be implemented by adding a `RetryPolicy` field to the client's config. This could be an optional field, which just defaults to no retries, when not specified by the user.

The new `Config` type could for example look like this:

```haskell
import Conrol.Retry (RetryPolicy)

data Config = Config
  }
  { configHost  :: ByteString
  , configPort  :: Word16
  , configToken :: Maybe ByteString
  , configRetryPolicy :: Maybe RetryPolicy  -- new field
  } deriving (Eq, Show)
```

### PR DESCRIPTION
**Pull request by DStranger**
Fixes #44

This is my attempt at adding retries to the Icepeak client.
I'm not really sure how to test this though, maybe you have suggestions?

### CODE REVIEW

File: client-haskell/src/Icepeak/Client.hs [hunk 1]
```diff
@@ -48,9 +53,14 @@ import qualified Network.HTTP.Client as HTTP
...
 import qualified Network.HTTP.Types.Header as Header
 import qualified Network.HTTP.Types.URI as URI
+import qualified Control.Retry as Retry
 
 -- | Config paired with HTTP manager for making requests.
-data Client = Client HTTP.Manager Config
+data Client m = Client 
+  { clientHttpManager :: HTTP.Manager
+  , clientConfig      :: Config
+  , clientRetryPolicy :: Maybe (RetryPolicyM m)
```
Comment (rightfold) [inline client-haskell/src/Icepeak/Client.hs]: Nothing and Just doNotRetry seem to encode the same information. I suggest making this field not Maybe, and exporting doNotRetry. That would simplify both the implementation and the usage of the library.

Comment (rightfold) [inline client-haskell/src/Icepeak/Client.hs]: Adding a new field here breaks the integration tests:

https://github.com/channable/icepeak/blob/fb30e95429010f1d46a29ca312724437da5ac596/server/integration-tests/stress_test.hs#L36

https://github.com/channable/icepeak/blob/fb30e95429010f1d46a29ca312724437da5ac596/server/integration-tests/auth_stress_test.hs#L39

Comment (DStranger) [reply client-haskell/src/Icepeak/Client.hs]: Yes, this seems like a good idea!

Comment (DStranger) [reply client-haskell/src/Icepeak/Client.hs]: Indeed, missed this.

**Comment by DStranger** [comment]
> I got two remarks but other than that: looks good! Very nice. 👍
Thanks!
I fixed those two things, had to force push after a rebase on master.

**Pull request merged**